### ISSUE TITLE
String.nsplit is not consistent with String.split_on_char

### ISSUE THREAD

**Issue by thizanne** [ocaml-batteries-team/batteries-included#845]
`String.nsplit` returns the empty list is the string parameter is empty. `split_on_char` never returns an empty list. 

Should we consider having them both behave the same, or is it best left as it is ? This should only be considered for next major release anyway, since it breaks some code.

**Comment by thizanne** [comment]
Actually `split_on_char` seems broken, it does return the empty list when applied to the empty string. See #846

**Comment by gasche** [comment]
`split_on_char` and `nsplit` have incompatible specifications, so unfortunately there is not much we can do to make them "coherent" before v3 -- but I believe that it's a better specification to never return an empty list. We should fix our `split_on_char` implementation now.

**Comment by rixed** [comment]
Not sure I understand this or if this is outdated already, but on batteries master I have:

```
utop # open String;;
utop # let s1 = split_on_char '|';;
val s1 : string -> string list = <fun>
utop # let s2 = nsplit ~by:"|";;
val s2 : string -> string list = <fun>
utop # s1 "" = s2 "";;
- : bool = true
utop # s1 "foo" = s2 "foo";;
- : bool = true
utop # s1 "foo|" = s2 "foo|";;
- : bool = true
utop # s1 "foo|bar" = s2 "foo|bar";;
- : bool = true
utop # s1 "|bar" = s2 "|bar";;
- : bool = true
utop # s1 "|" = s2 "|";;
- : bool = true
```

@thizanne what am I missing?

**Comment by thizanne** [comment]
Short answer in a hurry: #846

**Comment by rixed** [comment]
Ah I see.
So we want both BatString.split_on_char and BatString.nsplit to return `[""]` on an empty string.

**Comment by rixed** [comment]
One of my own program just got broken by this change, and as this change of API does not trigger any warning or anything, it took me an actual debugging session to figure out that the tests broke because of this change, and I'm the one who have implemented it. Also, I'm lucky that the bug has been caught by the tests. But I cannot help but think that my code is not the only one that will be subtly and silently broken.

I think it's better to change the API in a way that breaks compilation rather that runtime behavior.

Should we consider introducing a new function or a new optional parameter or anything like that to at least make sure users are aware of the change after an upgrade?

**Comment by UnixJunkie** [comment]
Maybe revert the change and resolve this issue as "won't fix".

**Comment by rixed** [comment]
We cannot merely `wontfix` this.
Two things are not acceptable, in my opinion:

1. That the behaviour of String.split_on_char changes at runtime depending on whether Batteries is opened or not
2. That the bahaviour of String.split_on_char or nsplit changes at runtime depending on the version of Batteries.

What is acceptable though is to break user programs at compile time.

Regarding 1. I think there is no better thing to do than to "fix" split_on_char so that it matches stdlib's, and consider it a bug (although that's not one). Users relying on split_on_char of an empty string to return the empty list are already in troubled water anyway, given the new stdlib function. I don't think there is much we can do to help them.

Regarding 2. I don't know. Changing split_on_char behaviour without changing that of nsplit is inelegant, but not wasting programers time by introducing subtle bugs sounds more important to me than elegance. I wish we could use the type system to break compilation of programs expecting an empty list from those functions :) (actually, there is a `warn_on_literal_pattern` attribute that would almost do that, but it applies to type definitions not to the return values of functions). Maybe we should leave nsplit as is, add a `deprecated` attribute, and have a new `split_on_string` or something like that for the new, more correct behaviour.

What do you think?

**Comment by UnixJunkie** [comment]
Ok, I agree with all your points. Maybe one idea that will break at compile time: we return an option type.
In case those functions are called on an empty string, we return None.
On a non empty string, we return Some x.
But then that would break the interface compatibility with the stdlib...

**Comment by UnixJunkie** [comment]
> We cannot merely `wontfix` this.
> Two things are not acceptable, in my opinion:
> 
>     1. That the behaviour of String.split_on_char changes at runtime depending on whether Batteries is opened or not

This is not cool. However, I feel that both functions are not meant to be called on an empty string.
So, only a corner case is changing.

>     2. That the bahaviour of String.split_on_char or nsplit changes at runtime depending on the version of Batteries.

Since this is for a major batteries release, people should not switched to it without carefully reading the breaking changes and maybe running extensive tests. They should use this new major version of batteries for new projects, not for their critical thing that is in production.

**Comment by rixed** [comment]
Implemented the changes, WDYT?

**Comment by UnixJunkie** [comment]
closed by PR #924
